lib/backend/test/atomicwrite_shim.go

```diff
@@ -0,0 +1,198 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package test
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/backend"
+)
+
+// RunBackendComplianceSuiteWithAtomicWriteShim runs the old backend compliance suite against the provided backend
+// with a shim that converts all calls to single-write methods (all write methods but DeleteRange) into calls to
+// AtomicWrite. This is done to ensure that the relationship between the conditional actions of AtomicWrite and the
+// single-write methods is well defined, and to improve overall coverage of AtomicWrite implementations via reuse.
+func RunBackendComplianceSuiteWithAtomicWriteShim(t *testing.T, newBackend AtomicWriteConstructor) {
+	RunBackendComplianceSuite(t, func(options ...ConstructionOption) (backend.Backend, clockwork.FakeClock, error) {
+		bk, clock, err := newBackend(options...)
+		if err != nil {
+			return nil, nil, trace.Wrap(err)
+		}
+
+		return atomicWriteShim{bk}, clock, nil
+	})
+}
+
+// atomciWriteShim reimplements all single-write backend methods as calls to AtomicWrite.
+type atomicWriteShim struct {
+	backend.AtomicWriteBackend
+}
+
+// Create creates item if it does not exist
+func (a atomicWriteShim) Create(ctx context.Context, i backend.Item) (*backend.Lease, error) {
+	rev, err := a.AtomicWrite(ctx, backend.ConditionalAction{
+		Key:       i.Key,
+		Condition: backend.NotExists(),
+		Action:    backend.Put(i),
+	})
+	if err != nil {
+		if errors.Is(err, backend.ErrConditionFailed) {
+			return nil, trace.AlreadyExists("already exists")
+		}
+		return nil, trace.Wrap(err)
+	}
+	return &backend.Lease{
+		Key:      i.Key,
+		Revision: rev,
+	}, nil
+}
+
+// Put puts value into backend (creates if it does not
+// exists, updates it otherwise)
+func (a atomicWriteShim) Put(ctx context.Context, i backend.Item) (*backend.Lease, error) {
+	rev, err := a.AtomicWrite(ctx, backend.ConditionalAction{
+		Key:       i.Key,
+		Condition: backend.Whatever(),
+		Action:    backend.Put(i),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &backend.Lease{
+		Key:      i.Key,
+		Revision: rev,
+	}, nil
+}
+
+// CompareAndSwap compares item with existing item
+// and replaces is with replaceWith item
+func (a atomicWriteShim) CompareAndSwap(ctx context.Context, expected backend.Item, replaceWith backend.Item) (*backend.Lease, error) {
```

Comment:
Do we guarantee that `Put("k", "v")` followed by some concurrent `CompareAndSwap("k", "v", "v")`s and `Put("k", "v")`s will all succeed? If so, this needs a retry loop (cfr. `(*lib/backend/firestore.Backend).CompareAndSwap`).